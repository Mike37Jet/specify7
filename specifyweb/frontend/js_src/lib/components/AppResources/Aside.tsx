import React from 'react';

import type { AppResourceFilters as AppResourceFiltersType } from './filtersHelpers';
import { buildAppResourceConformation, getAppResourceMode } from './helpers';
import type {
  SpAppResource,
  SpViewSetObj as SpViewSetObject,
} from '../DataModel/types';
import type { SerializedResource } from '../DataModel/helpers';
import { removeItem, replaceItem } from '../../utils/utils';
import { adminText } from '../../localization/admin';
import { commonText } from '../../localization/common';
import { hasToolPermission } from '../Permissions/helpers';
import type { RA } from '../../utils/types';
import { AppResourceIcon } from './EditorComponents';
import { AppResourcesFilters, useFilteredAppResources } from './Filters';
import type { AppResources, AppResourcesTree } from './hooks';
import { useAppResourceCount, useResourcesTree } from './hooks';
import { ActiveLink } from '../Molecules';
import { icons } from '../Atoms/Icons';
import { useCachedState } from '../../hooks/useCachedState';
import { useErrorContext } from '../../hooks/useErrorContext';
import { Ul } from '../Atoms';
import { className } from '../Atoms/className';
import { Button } from '../Atoms/Button';
import { Link } from '../Atoms/Link';
import {useId} from '../../hooks/useId';

export function AppResourcesAside({
  resources: initialResources,
  isReadOnly,
  initialFilters,
  onOpen: handleOpen,
}: {
  readonly resources: AppResources;
  readonly isReadOnly: boolean;
  readonly initialFilters?: AppResourceFiltersType;
  readonly onOpen?: (
    resource: SerializedResource<SpAppResource | SpViewSetObject>
  ) => void;
}): JSX.Element {
  const [conformations = [], setConformations] = useCachedState(
    'appResources',
    'conformation'
  );
  const resources = useFilteredAppResources(initialResources, initialFilters);
  const resourcesTree = useResourcesTree(resources);
  useErrorContext('appResourcesTree', resourcesTree);

  return (
    <aside className={className.containerBase}>
      <Ul className="flex flex-1 flex-col gap-1 overflow-auto" role="tree">
        {resourcesTree.map((resources) => (
          <TreeItem
            conformations={conformations}
            isReadOnly={isReadOnly}
            key={resources.key}
            resourcesTree={resources}
            onFold={setConformations}
            onOpen={handleOpen}
          />
        ))}
      </Ul>
      <div className="flex flex-wrap gap-2">
        <AppResourcesFilters initialResources={initialResources} />
        <AppResourcesExpand
          resourcesTree={resourcesTree}
          onChange={setConformations}
        />
      </div>
    </aside>
  );
}

export type AppResourcesConformation = {
  readonly key: string;
  readonly children: RA<AppResourcesConformation>;
};

function AppResourcesExpand({
  resourcesTree,
  onChange: handleChange,
}: {
  readonly resourcesTree: AppResourcesTree;
  readonly onChange: (conformation: RA<AppResourcesConformation>) => void;
}): JSX.Element {
  return (
    <>
      <Button.Blue
        onClick={(): void =>
          handleChange(buildAppResourceConformation(resourcesTree))
        }
      >
        {commonText('expandAll')}
      </Button.Blue>
      <Button.Blue onClick={(): void => handleChange([])}>
        {commonText('collapseAll')}
      </Button.Blue>
    </>
  );
}

function TreeItem({
  resourcesTree,
  conformations,
  isReadOnly,
  onFold: handleFold,
  onOpen: handleOpen,
}: {
  readonly resourcesTree: AppResourcesTree[number];
  readonly conformations: RA<AppResourcesConformation>;
  readonly isReadOnly: boolean;
  readonly onFold: (conformations: RA<AppResourcesConformation>) => void;
  readonly onOpen:
    | ((resource: SerializedResource<SpAppResource | SpViewSetObject>) => void)
    | undefined;
}): JSX.Element {
  const { label, key, subCategories } = resourcesTree;
  const conformationIndex = conformations.findIndex(
    (conformation) => conformation.key === key
  );
  const isExpanded = conformationIndex !== -1;
  const conformation = conformations[conformationIndex] ?? {
    key,
    children: [],
  };
  const id = useId('app-resources-tree');
  const count = useAppResourceCount(resourcesTree);
  return (
    <li
      aria-expanded={isExpanded}
      aria-labelledby={id('label')}
      className="flex flex-col gap-2"
      id={id('li')}
      role="treeitem"
    >
      <Button.LikeLink
        aria-controls={id('li')}
        className="font-bold"
        id={id('label')}
        onClick={(): void =>
          handleFold(
            conformationIndex === -1
              ? [...conformations, conformation]
              : removeItem(conformations, conformationIndex)
          )
        }
      >
        {label}
        <span className="text-neutral-500">{` (${count})`}</span>
      </Button.LikeLink>
      {isExpanded && (
        <>
          <TreeItemResources
            isReadOnly={isReadOnly}
            resourcesTree={resourcesTree}
            onOpen={handleOpen}
          />
          {subCategories.length > 0 && (
            <Ul
              aria-label={adminText('subCategories')}
              className="flex flex-col gap-2 pl-4"
              role="group"
            >
              {subCategories.map((resources) => (
                <TreeItem
                  conformations={conformation.children}
                  isReadOnly={isReadOnly}
                  key={resources.key}
                  resourcesTree={resources}
                  onFold={(newConformation): void =>
                    handleFold(
                      mutateConformation(conformations, key, newConformation)
                    )
                  }
                  onOpen={handleOpen}
                />
              ))}
            </Ul>
          )}
        </>
      )}
    </li>
  );
}

function TreeItemResources({
  resourcesTree,
  isReadOnly,
  onOpen: handleOpen,
}: {
  readonly resourcesTree: AppResourcesTree[number];
  readonly isReadOnly: boolean;
  readonly onOpen:
    | ((resource: SerializedResource<SpAppResource | SpViewSetObject>) => void)
    | undefined;
}): JSX.Element | null {
  const { appResources, viewSets, directory, key } = resourcesTree;
  const resources = [...appResources, ...viewSets];
  const canCreate = hasToolPermission('resources', 'create') && !isReadOnly;
  return typeof directory === 'object' &&
    (resources.length > 0 || canCreate) ? (
    <Ul aria-label={adminText('resources')} className="pl-4" role="group">
      {resources.map((resource, index) => (
        <li key={index}>
          <ResourceItem resource={resource} onOpen={handleOpen} />
        </li>
      ))}
      {canCreate && (
        <li>
          <Link.Default href={`/specify/resources/create/${key}/`}>
            <span className={className.dataEntryAdd}>{icons.plus}</span>
            {adminText('addResource')}
          </Link.Default>
        </li>
      )}
    </Ul>
  ) : null;
}

function mutateConformation(
  conformations: RA<AppResourcesConformation>,
  key: string,
  childConformation: RA<AppResourcesConformation>
): RA<AppResourcesConformation> {
  const conformationIndex = conformations.findIndex(
    (conformation) => conformation.key === key
  );
  return replaceItem(conformations, conformationIndex, {
    key,
    children: childConformation,
  });
}

function ResourceItem({
  resource,
  onOpen: handleOpen,
}: {
  readonly resource: SerializedResource<SpAppResource | SpViewSetObject>;
  readonly onOpen:
    | ((resource: SerializedResource<SpAppResource | SpViewSetObject>) => void)
    | undefined;
}): JSX.Element {
  const url =
    getAppResourceMode(resource) === 'appResources'
      ? `/specify/resources/app-resource/${resource.id}/`
      : `/specify/resources/view-set/${resource.id}/`;
  return (
    <ActiveLink
      className="!text-neutral-500 hover:!text-brand-300"
      href={url}
      onClick={
        typeof handleOpen === 'function'
          ? (event): void => {
              event.preventDefault();
              handleOpen(resource);
            }
          : undefined
      }
    >
      <AppResourceIcon resource={resource} />
      {resource.name || commonText('nullInline')}
    </ActiveLink>
  );
}
