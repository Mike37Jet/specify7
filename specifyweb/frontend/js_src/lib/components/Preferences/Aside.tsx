import React from 'react';
import { useLocation } from 'react-router';
import { useNavigate } from 'react-router-dom';

import { listen } from '../../utils/events';
import { f } from '../../utils/functools';
import type { GetSet, WritableArray } from '../../utils/types';
import { filterArray } from '../../utils/types';
import { Link } from '../Atoms/Link';
import { locationToState } from '../Router/RouterState';
import { usePrefDefinitions } from './index';

/** Update the active category on the sidebar as user scrolls */
export function useActiveCategory(): {
  readonly activeCategory: number;
  readonly forwardRefs: (index: number, element: HTMLElement | null) => void;
  readonly containerRef: React.RefCallback<HTMLDivElement | null>;
} {
  const [activeCategory, setActiveCategory] = React.useState<number>(0);
  const observer = React.useRef<IntersectionObserver | undefined>(undefined);
  const references = React.useRef<WritableArray<HTMLElement | undefined>>([]);
  React.useEffect(() => () => observer.current?.disconnect(), []);

  // eslint-disable-next-line functional/prefer-readonly-type
  const intersecting = React.useRef<Set<number>>(new Set());

  function handleObserved({
    isIntersecting,
    target,
  }: IntersectionObserverEntry): void {
    const index = references.current.indexOf(target as HTMLElement);
    intersecting.current[isIntersecting ? 'add' : 'delete'](index);
    const intersection = f.min(...Array.from(intersecting.current)) ?? 0;
    setActiveCategory(intersection);
  }

  return {
    activeCategory,
    forwardRefs: React.useCallback((index, element) => {
      const oldElement = references.current[index];
      if (typeof oldElement === 'object')
        observer.current?.unobserve(oldElement);
      references.current[index] = element ?? undefined;
      if (element !== null) observer?.current?.observe(element);
    }, []),
    containerRef: React.useCallback((container): void => {
      observer.current?.disconnect();

      observer.current = new IntersectionObserver(
        (entries) => entries.map(handleObserved),
        {
          root: container,
          rootMargin: '-200px 0px -100px 0px',
          threshold: 0,
        }
      );
      /*
       * Since React 18, apps running in strict mode in development are mounted
       * followed immediately by an unmount and then mount again. This causes
       * observer not to fire. Can be fixed by either
       * running React in non-strict mode (bad idea), or wrapping the following
       * in setTimeout(()=>..., 0);
       * More info:
       * https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-strict-mode
       */
      setTimeout(
        () =>
          filterArray(references.current).forEach((value) =>
            observer.current?.observe(value)
          ),
        0
      );
    }, []),
  };
}

export function PreferencesAside({
  id,
  activeCategory,
}: {
  readonly id: (prefix: string) => string;
  readonly activeCategory: number;
}): JSX.Element {
  const definitions = usePrefDefinitions();
  const navigate = useNavigate();
  const location = useLocation();
  const state = locationToState(location, 'BackgroundLocation');
  const isInOverlay = typeof state === 'object';
  React.useEffect(
    () =>
      isInOverlay
        ? undefined
        : navigate(
            `/specify/user-preferences/#${id(definitions[activeCategory][0])}`,
            {
              replace: true,
            }
          ),
    [isInOverlay, definitions, activeCategory, id]
  );

  const [freezeCategory, setFreezeCategory] = useFrozenCategory();
  const currentIndex = freezeCategory ?? activeCategory;

  return (
    <aside
      className={`
        top-0 flex min-w-fit flex-1 flex-col divide-y-4 divide-[color:var(--form-background)]
        overflow-y-auto md:sticky
      `}
    >
      {definitions.map(([category, { title }], index) => (
        <Link.Gray
          aria-current={currentIndex === index ? 'page' : undefined}
          href={`#${id(category)}`}
          key={category}
          onClick={(): void => setFreezeCategory(index)}
        >
          {title}
        </Link.Gray>
      ))}
    </aside>
  );
}

/**
 * Clicking on a category scrolls that category to the top.
 * The active category is determined based on which category is in the middle.
 * Thus, clicking on a category might not make it active.
 * This hack temporary makes the clicked category active, until user starts to
 * scroll away
 */
function useFrozenCategory(): GetSet<number | undefined> {
  const [freezeCategory, setFreezeCategory] = React.useState<
    number | undefined
  >(undefined);
  React.useEffect(
    () =>
      typeof freezeCategory === 'number'
        ? listen(
            document.body,
            'resize',
            () => setFreezeCategory(undefined),
            true
          )
        : undefined,
    [freezeCategory]
  );
  React.useEffect(
    () =>
      typeof freezeCategory === 'number'
        ? listen(
            document.body,
            'scroll',
            () => {
              if (ignoreScroll.current) ignoreScroll.current = false;
              else setFreezeCategory(undefined);
            },
            true
          )
        : undefined,
    [freezeCategory]
  );

  /*
   * Clicking on a category scrolls to it. Need to ignore that initial scroll
   * and only listen for subsequent scrolls
   */
  const ignoreScroll = React.useRef<boolean>(false);

  return [
    freezeCategory,
    React.useCallback((value) => {
      ignoreScroll.current = typeof value === 'number';
      setFreezeCategory(value);
    }, []),
  ];
}
