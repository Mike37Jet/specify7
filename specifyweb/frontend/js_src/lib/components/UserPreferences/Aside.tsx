import React from 'react';
import { useNavigate } from 'react-router-dom';

import type { WritableArray } from '../../utils/types';
import { filterArray } from '../../utils/types';
import { Link } from '../Atoms/Link';
import { usePrefDefinitions } from './index';
import { f } from '../../utils/functools';

/** Update the active category on the sidebar as user scrolls */
export function useActiveCategory(): {
  readonly activeCategory: number;
  readonly forwardRefs: (index: number, element: HTMLElement | null) => void;
  readonly containerRef: React.RefCallback<HTMLDivElement | null>;
} {
  const [activeCategory, setActiveCategory] = React.useState<number>(0);
  const observer = React.useRef<IntersectionObserver | undefined>(undefined);
  const references = React.useRef<WritableArray<HTMLElement | undefined>>([]);
  React.useEffect(() => () => observer.current?.disconnect(), []);

  // eslint-disable-next-line functional/prefer-readonly-type
  const intersecting = React.useRef<Set<number>>(new Set());

  function handleObserved({
    isIntersecting,
    target,
  }: IntersectionObserverEntry): void {
    const index = references.current.indexOf(target as HTMLElement);
    intersecting.current[isIntersecting ? 'add' : 'delete'](index);
    const intersection = f.min(...Array.from(intersecting.current)) ?? 0;
    setActiveCategory(intersection);
  }

  return {
    activeCategory,
    forwardRefs: React.useCallback((index, element) => {
      const oldElement = references.current[index];
      if (typeof oldElement === 'object')
        observer.current?.unobserve(oldElement);
      references.current[index] = element ?? undefined;
      if (element !== null) observer?.current?.observe(element);
    }, []),
    containerRef: React.useCallback((container): void => {
      observer.current?.disconnect();

      observer.current = new IntersectionObserver(
        (entries) => entries.map(handleObserved),
        {
          root: container,
          rootMargin: '-200px 0px -100px 0px',
          threshold: 0,
        }
      );
      /*
       * Since React 18, apps running in strict mode are mounted followed
       * immediately by an unmount and the mount again when running in
       * development. This causes observer not to fire. Can be fixed by either
       * running React in non-strict mode (bad idea), or wrapping the following
       * in setTimeout(()=>..., 0);
       * More info:
       * https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-strict-mode
       */
      setTimeout(
        () =>
          filterArray(references.current).forEach((value) =>
            observer.current?.observe(value)
          ),
        0
      );
    }, []),
  };
}

export function PreferencesAside({
  id,
  activeCategory,
}: {
  readonly id: (prefix: string) => string;
  readonly activeCategory: number;
}): JSX.Element {
  const definitions = usePrefDefinitions();
  const navigate = useNavigate();
  React.useEffect(
    () =>
      navigate(
        `/specify/user-preferences/#${id(definitions[activeCategory][0])}`,
        {
          replace: true,
        }
      ),
    [definitions, activeCategory, id]
  );
  return (
    <aside
      className={`
        top-0 flex min-w-fit flex-1 flex-col divide-y-4 divide-[color:var(--form-background)]
        md:sticky
      `}
    >
      {definitions.map(([category, { title }], index) => (
        <Link.Gray
          aria-current={activeCategory === index ? 'page' : undefined}
          href={`#${id(category)}`}
          key={category}
        >
          {title}
        </Link.Gray>
      ))}
    </aside>
  );
}
