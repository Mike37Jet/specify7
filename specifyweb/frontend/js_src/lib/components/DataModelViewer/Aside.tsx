import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { locationToState } from '../Router/RouterState';
import { getTables } from './helpers';
import { useFrozenCategory } from '../UserPreferences/Aside';
import { Link } from '../Atoms/Link';

export function DataModelAside({
  activeCategory,
}: {
  readonly activeCategory: number | undefined;
}): JSX.Element {
  const tables = React.useMemo(getTables, []);
  const [freezeCategory, setFreezeCategory] = useFrozenCategory();
  const currentIndex = freezeCategory ?? activeCategory;
  const navigate = useNavigate();
  const location = useLocation();
  const state = locationToState(location, 'BackgroundLocation');
  const isInOverlay = typeof state === 'object';

  React.useEffect(
    () =>
      isInOverlay || activeCategory === undefined
        ? undefined
        : navigate(`/specify/data-model/#${tables[activeCategory].name[0]}`, {
            replace: true,
          }),
    [isInOverlay, tables, activeCategory]
  );

  return (
    <aside
      className={`
        left-0 hidden min-w-fit flex-1 flex-col divide-y-4
        divide-[color:var(--form-background)] overflow-y-auto md:flex
      `}
    >
      {tables.map(({ name: [tableName, jsxName] }, index) => (
        <Link.Gray
          aria-current={currentIndex === index ? 'page' : undefined}
          className="!justify-start"
          href={`#${tableName}`}
          key={index}
          onClick={(): void => setFreezeCategory(index)}
        >
          {jsxName}
        </Link.Gray>
      ))}
    </aside>
  );
}
