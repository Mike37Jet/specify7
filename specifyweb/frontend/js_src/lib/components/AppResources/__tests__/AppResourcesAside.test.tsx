import { render } from "@testing-library/react"
import { AppResourcesAside } from "../Aside"
import React from "react"
import { staticAppResources } from "./staticAppResources"

describe("AppResourcesAside", () => {

    test("simple no resources case", () => {

        const onOpen = jest.fn();
        const setConformations = jest.fn();

        render(<AppResourcesAside
            resources={staticAppResources}
            isEmbedded
            onOpen={onOpen}
            filters={undefined}
            conformations={[[], setConformations]}
        />);

    })
})