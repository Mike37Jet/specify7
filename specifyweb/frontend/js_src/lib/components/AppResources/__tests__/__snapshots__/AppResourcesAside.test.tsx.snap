// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppResourcesAside (expanded case) expanded case 1`] = `
<DocumentFragment>
  <aside
    class="
        !gap-2 sm:overflow-visible
        flex flex-col gap-4 overflow-scroll
  overflow-x-auto [overflow-y:overlay] [scrollbar-gutter:auto] 
      "
  >
    <ul
      class=" flex flex-1 flex-col gap-1 overflow-auto"
      role="tree"
    >
      <li
        aria-expanded="false"
        aria-labelledby="app-resources-tree-4-label"
        class="flex flex-col gap-2"
        id="app-resources-tree-4-li"
        role="treeitem"
      >
        <button
          aria-controls="app-resources-tree-4-li"
          class="link inline text-left font-bold"
          id="app-resources-tree-4-label"
          type="button"
        >
          Global Resources 
          <span
            class="pl-2 text-neutral-500"
          >
            (2)
          </span>
        </button>
      </li>
      <li
        aria-expanded="true"
        aria-labelledby="app-resources-tree-5-label"
        class="flex flex-col gap-2"
        id="app-resources-tree-5-li"
        role="treeitem"
      >
        <button
          aria-controls="app-resources-tree-5-li"
          class="link inline text-left font-bold"
          id="app-resources-tree-5-label"
          type="button"
        >
          Discipline Resources 
          <span
            class="pl-2 text-neutral-500"
          >
            (4)
          </span>
        </button>
        <ul
          aria-label="Sub-categories"
          class=" flex flex-col gap-2 pl-4"
          role="group"
        >
          <li
            aria-expanded="false"
            aria-labelledby="app-resources-tree-6-label"
            class="flex flex-col gap-2"
            id="app-resources-tree-6-li"
            role="treeitem"
          >
            <button
              aria-controls="app-resources-tree-6-li"
              class="link inline text-left font-bold"
              id="app-resources-tree-6-label"
              type="button"
            >
              Botany 
              <span
                class="pl-2 text-neutral-500"
              >
                (4)
              </span>
            </button>
          </li>
        </ul>
      </li>
    </ul>
    <div
      class="flex flex-wrap gap-2"
    >
      <button
        class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white grow"
        type="button"
      >
        Expand All
      </button>
      <button
        class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white grow"
        type="button"
      >
        Collapse All
      </button>
    </div>
  </aside>
</DocumentFragment>
`;

exports[`AppResourcesAside (simple no conformation case) simple no conformation case 1`] = `
<DocumentFragment>
  <aside
    class="
        !gap-2 sm:overflow-visible
        flex flex-col gap-4 overflow-scroll
  overflow-x-auto [overflow-y:overlay] [scrollbar-gutter:auto] 
      "
  >
    <ul
      class=" flex flex-1 flex-col gap-1 overflow-auto"
      role="tree"
    >
      <li
        aria-expanded="false"
        aria-labelledby="app-resources-tree-0-label"
        class="flex flex-col gap-2"
        id="app-resources-tree-0-li"
        role="treeitem"
      >
        <button
          aria-controls="app-resources-tree-0-li"
          class="link inline text-left font-bold"
          id="app-resources-tree-0-label"
          type="button"
        >
          Global Resources 
          <span
            class="pl-2 text-neutral-500"
          >
            (2)
          </span>
        </button>
      </li>
      <li
        aria-expanded="false"
        aria-labelledby="app-resources-tree-1-label"
        class="flex flex-col gap-2"
        id="app-resources-tree-1-li"
        role="treeitem"
      >
        <button
          aria-controls="app-resources-tree-1-li"
          class="link inline text-left font-bold"
          id="app-resources-tree-1-label"
          type="button"
        >
          Discipline Resources 
          <span
            class="pl-2 text-neutral-500"
          >
            (4)
          </span>
        </button>
      </li>
    </ul>
    <div
      class="flex flex-wrap gap-2"
    >
      <button
        class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white grow"
        type="button"
      >
        Expand All
      </button>
      <button
        class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white grow"
        type="button"
      >
        Collapse All
      </button>
    </div>
  </aside>
</DocumentFragment>
`;
